Update nuget.exe push to Azdo Artifact example

erdembayar opened the pull request:
Updates: https://docs.microsoft.com/en-us/azure/devops/artifacts/nuget/publish?view=azure-devops
Related to: https://github.com/NuGet/docs.microsoft.com-nuget/issues/2785
Description: Updates how to authenticate on local and hosted build and added examples.

PRMerger9 commented:
@erdembayar : Thanks for your contribution! The author(s) have been notified to review your proposed change.

PRMerger13 commented:
@erdembayar : Thanks for your contribution! The author(s) have been notified to review your proposed change.

PRMerger16 commented:
@erdembayar : Thanks for your contribution! The author(s) have been notified to review your proposed change.

PRMerger12 commented:
@erdembayar : Thanks for your contribution! The author(s) have been notified to review your proposed change.

Court72 commented:
@ramiMSFT 

Can you review the proposed changes? If you approve the changes, they should be moved to the private repo for staging and validation. 

When the changes are ready for publication, add a `#sign-off` comment to signal that the PR is ready for the review team to merge. 

#label:"aq-pr-triaged"
@MicrosoftDocs/public-repo-pr-review-team

zivkan commented on docs/artifacts/includes/nuget/nuget-publish-endpoint.md:
```diff
@@ -18,14 +18,38 @@ ms.date: 02/18/2022
     :::image type="content" source="../../media/nuget-connect-feed.png" alt-text="NuGet.exe feed connection":::
 
-1. If this is the first time using Azure Artifacts with Nuget.exe, select **Get the tools** button and follow the instructions to install the prerequisites.
+1. When using a local development environment, you must have the [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) installed which provides authentication when pushing or downloading packages (see examples section).
```
```suggestion
1. When using a local development environment, it is recommended to have the [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) installed which provides authentication when pushing or downloading packages (see examples section).
```

Technically the PAT can be saved in a nuget.config as a package source credential, it's just better to use the credential provider.

PRMerger16 commented:
@erdembayar : Thanks for your contribution! The author(s) have been notified to review your proposed change.

PRMerger14 commented:
@erdembayar : Thanks for your contribution! The author(s) have been notified to review your proposed change.

PRMerger6 commented:
@erdembayar : Thanks for your contribution! The author(s) have been notified to review your proposed change.

PRMerger14 commented:
@erdembayar : Thanks for your contribution! The author(s) have been notified to review your proposed change.

PRMerger20 commented:
@erdembayar : Thanks for your contribution! The author(s) have been notified to review your proposed change.

erdembayar commented:
@zivkan 
Could you review 1 more time? I added examples for pushing from `Github action`  to Azure and other hosted builds to Azure using PAT.

erdembayar commented on docs/artifacts/includes/nuget/nuget-publish-endpoint.md:
```diff
@@ -18,14 +18,78 @@ ms.date: 02/18/2022
+
+1. For Github actions builds that interact with Azure Artifacts feeds, you may use [Github setup dotnet action](https://github.com/actions/setup-dotnet) for the authentication(see examples section).
+
+1. For other hosted builds that interact with Azure Artifacts feeds, you may use [Azure PAT]([url](https://docs.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=Windows)) saved in `nuget.config` file for the authentication(see examples section), but please exercise caution because it may leak your PAT as part of build enviro ...
```
I added `other hosted builds` section because there're other hosting such as Teamcity, Jenkins etc which may don't support [NuGet Authenticate](https://docs.microsoft.com/azure/devops/pipelines/tasks/package/nuget-authenticate) task, but still want to publish nupkg package to Azure Artifacts.

nkolev92 commented on docs/artifacts/includes/nuget/nuget-publish-endpoint.md:
````diff
@@ -18,14 +18,78 @@ ms.date: 02/18/2022
+- In the example below, pushing "MyPackage" version "5.0.2" to an Azure Artifacts feed from a local development environment, the API key "AZ" is only used as a placeholder. An Azure Artifacts feed still requires the [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) to be installed to properly authenticate.
+
+```dotnetcli
+nuget push MyPackage.5.0.2.nupkg -src https://pkgs.dev.azure.com/{organization}/{project}/_packaging/{feed}/nuget/v3/index.json AZ
````
I think you're missing an api key option here.

nkolev92 replied:
I see. As long as it works I guess. :D

nkolev92 commented on docs/artifacts/includes/nuget/nuget-publish-endpoint.md:
```diff
@@ -18,14 +18,78 @@ ms.date: 02/18/2022
+
+## Examples
+
+- In the example below, pushing "MyPackage" version "5.0.2" to an Azure Artifacts feed from a local development environment, the API key "AZ" is only used as a placeholder. An Azure Artifacts feed still requires the [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) to be installed to properly authenticate.
```
Another idea is to put `ignored` as the API-key value. 

Makes it very explicit that the vlaue is irrelevant.

erdembayar replied:
Good input. Still `AZ` is short, also subtly express it's Azure only quirk. Let's keep it as it's for now, I'm welcome to other ideas.

nkolev92 replied:
I think you might have misunderstood me. 

If you put the value in the example, it'll be obvious what is going on and you wouldn't need the sentence explaining the placeholder.

erdembayar replied:
Removed the `(it can any value)`

nkolev92 commented on docs/artifacts/includes/nuget/nuget-publish-endpoint.md:
```diff
@@ -18,14 +18,78 @@ ms.date: 02/18/2022
+        nuGetServiceConnections: MyServiceConnection_ExternalServer
+        
+    - script: |
+        dotnet build mypackage/mypackage.csproj -c Release
```
nit: I'd just include the push example. 

It's a slippery slope once you add all the steps.

nkolev92 replied:
Running `dotnet build`, `dotnet pack`  in the same script is not the best idea either as they are implicit. 

You could always just run `dotnet pack`, but I still think it isn't necessary. 

Curious what others think.

zivkan replied:
I agree. Many developers will want restore, build, pack, and push in different steps, so that if something fails, it's much more obvious what step it failed at, before even reading the logs.

Also, I find it odd that build and pack is done with the `dotnet` cli, but then push with nuget.exe.  If you're going to build with the dotnet CLI, you might as well also push with it.

Honestly, I don't think there should be separate pages for nuget.exe and dotnet cli. I think there should just be a single "push from cli" page, and it should only use the dotnet cli.  Let's not encourage customers to use nuget.exe.

erdembayar replied:
Sure, I removed unnecessary `dotnet build` from all examples.

erdembayar replied:
> Honestly, I don't think there should be separate pages for nuget.exe and dotnet cli. I think there should just be a single "push from cli" page, and it should only use the dotnet cli. Let's not encourage customers to use nuget.exe.

If I remember correctly there're some edge cases like mono which still use nuget.exe, so keep it separate for now. Removing a page easier thing to do next time, so let's deal next time.

nkolev92 commented on docs/artifacts/includes/nuget/nuget-publish-endpoint.md:
````diff
@@ -18,14 +18,78 @@ ms.date: 02/18/2022
+      displayName: "Pack and push"          
+  ```
+ 
+- In the example below, pushing "MyPackage" version "1.1.3" to an Azure Artifacts feed from an Github action build, the API key "AZ" is only used as a placeholder. An Azure Artifacts feed still requires your pipeline to use the [Github setup dotnet action](https://github.com/actions/setup-dotnet) for the authentication.
````
> In the example below, pushing "MyPackage" version "1.1.3" to an Azure Artifacts feed from an Github action build, the API key "AZ" is only used as a placeholder

This reads a bit funny to me. The subject is missing. 

When writing examples, you are also better off putting a header rather repeating things like `In the example below`.
Try to keep things concise as well.  

An idea would be: 

```
### Pushing from GitHub Actions

When pushing from GitHub actions your pipeline should use the [Github setup dotnet action](https://github.com/actions/setup-dotnet) for the authentication.

---example details---
```

zivkan replied:
> Try to keep things concise as well.

I agree. Have a look at the examples on other pages. Are they "complete" examples, showing common steps needed to get the point where it can demonstrate the command this page is about, or are they short, showing only the minimum for the current page's concept?

erdembayar replied:
I improved according to your recommendation, please check now.

nkolev92 commented on docs/artifacts/includes/nuget/nuget-publish-endpoint.md:
```diff
@@ -18,14 +18,78 @@ ms.date: 02/18/2022
+    
+    - name: Pack and push
+      run: |
+          dotnet build mypackage/mypackage.csproj -c Release
```
similar feedback as the previous one.

nkolev92 reviewed (commented):
I like the examples. It does feel a bit wordy right now. I think we can make it more concise and easier to follow.

zivkan commented:
@Court72 @ramiMSFT other docs repos have the docs bot build and post a preview link so we can manually check the rendered docs. Is that possible here?

zivkan commented on docs/artifacts/includes/nuget/nuget-publish-endpoint.md:
````diff
@@ -18,14 +18,78 @@ ms.date: 02/18/2022
+
+- In the example below, pushing "MyPackage" version "5.0.2" to an Azure Artifacts feed from a local development environment, the API key "AZ" is only used as a placeholder. An Azure Artifacts feed still requires the [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) to be installed to properly authenticate.
+
+```dotnetcli
````
I think `Command` is correct. At least, that's what they use in other docs pages: https://github.com/MicrosoftDocs/azure-devops-docs/blame/main/docs/artifacts/get-started-nuget.md#L83-L92

zivkan commented:
Looks like there's an error on a different page: https://github.com/MicrosoftDocs/azure-devops-docs/blame/main/docs/artifacts/nuget/nuget-exe.md#L111-L116

Might be worthwhile fixing it in this PR as well.

PRMerger12 commented:
@erdembayar : Thanks for your contribution! The author(s) have been notified to review your proposed change.

erdembayar commented on docs/artifacts/includes/nuget/nuget-publish-endpoint.md:
```diff
@@ -18,14 +18,78 @@ ms.date: 02/18/2022
     :::image type="content" source="../../media/nuget-connect-feed.png" alt-text="NuGet.exe feed connection":::
 
-1. If this is the first time using Azure Artifacts with Nuget.exe, select **Get the tools** button and follow the instructions to install the prerequisites.
+1. When using a local development environment, it is recommended to have the [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) installed which provides authentication when pushing or downloading packages (see examples section).
```
Sure, fixed.

PRMerger5 commented:
@erdembayar : Thanks for your contribution! The author(s) have been notified to review your proposed change.

PRMerger19 commented:
@erdembayar : Thanks for your contribution! The author(s) have been notified to review your proposed change.

PRMerger7 commented:
@erdembayar : Thanks for your contribution! The author(s) have been notified to review your proposed change.

PRMerger-2 commented:
@erdembayar : Thanks for your contribution! The author(s) have been notified to review your proposed change.

PRMerger7 commented:
@erdembayar : Thanks for your contribution! The author(s) have been notified to review your proposed change.

erdembayar commented:
> 
> Hi @zivkan - You could move these commits into the private repo where staging and validation are enabled. The steps are in the [Contributor guide](https://review.docs.microsoft.com/en-us/help/contribute/contribute-how-to-manage-act-guidelines-public-pr?branch=main#pull-public-commits-to-the-private-repo-for-updates-staging-publication).

@nkolev92 @zivkan 

@ramiMSFT and I created latest preview builds, please review, let me know if you have any comments.
We made some adjustments in private branch so it may look different from my current branch
 
**nuget.exe push example**: https://review.docs.microsoft.com/en-us/azure/devops/artifacts/nuget/publish?view=azure-devops&branch=pr-en-us-4782
[Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) requirement added into [Prerequisites section](https://review.docs.microsoft.com/en-us/azure/devops/artifacts/nuget/publish?view=azure-devops&branch=pr-en-us-4782#prerequisites).
We added push from other [external sources (for teamcity, jenkins etc ..) section](https://review.docs.microsoft.com/en-us/azure/devops/artifacts/nuget/publish?view=azure-devops&branch=pr-en-us-4782#publish-packages-from-external-sources) at bottom.

**Push from pipeline example**: We added script example as alternative to nuget.command task in pipeline example: https://review.docs.microsoft.com/en-us/azure/devops/pipelines/artifacts/nuget?toc=%2Fazure%2Fdevops%2Fartifacts%2Ftoc.json&bc=%2Fazure%2Fdevops%2Fartifacts%2Fbreadcrumb%2Ftoc.json&view=azure-devops&branch=pr-en-us-4782&tabs=yaml#publish-a-package

![image](https://user-images.githubusercontent.com/8766776/175612861-fcafc200-bfeb-4477-a7e6-cd198c623f27.png)

We'll create another `dotnet` PR separately.

nkolev92 reviewed (approved):
Looks good.

Mostly cleaning up some prepositions and one link that you may want to consider fixing.

erdembayar commented on docs/artifacts/includes/nuget/nuget-publish-endpoint.md:
```diff
@@ -18,14 +18,69 @@ ms.date: 02/18/2022
+
+### Pushing from a local development environment
+
+When pushing from local development environment to Azure Artifacts, it is recommended to have [Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) installed for the authentication. Please note here the API key "AZ" is only used as a placeholder.
```
Sorry, my current PR doesn't reflect latest changes in private branch we built for preview. I should have been more explicit.

We made some adjustments in private branch so it may look different from my current branch
Could you look at private branch instead of changes in this PR?

**nuget.exe push example**: https://review.docs.microsoft.com/en-us/azure/devops/artifacts/nuget/publish?view=azure-devops&branch=pr-en-us-4782
[Azure Artifacts Credential Provider](https://github.com/microsoft/artifacts-credprovider) requirement added into [Prerequisites section](https://review.docs.microsoft.com/en-us/azure/devops/artifacts/nuget/publish?view=azure-devops&branch=pr-en-us-4782#prerequisites).
We added push from other [external sources (for teamcity, jenkins etc ..) section](https://review.docs.microsoft.com/en-us/azure/devops/artifacts/nuget/publish?view=azure-devops&branch=pr-en-us-4782#publish-packages-from-external-sources) at bottom.

**Push from pipeline example**: We added script example as alternative to nuget.command task in pipeline example: https://review.docs.microsoft.com/en-us/azure/devops/pipelines/artifacts/nuget?toc=%2Fazure%2Fdevops%2Fartifacts%2Ftoc.json&bc=%2Fazure%2Fdevops%2Fartifacts%2Fbreadcrumb%2Ftoc.json&view=azure-devops&branch=pr-en-us-4782&tabs=yaml#publish-a-package

nkolev92 replied:
Can we review that PR instead?

nkolev92 replied:
I don't have access to view the preview.

erdembayar replied:
Private PR link is: https://github.com/MicrosoftDocs/azure-devops-docs-pr/pull/4782
If you get 404 error then you need to join `MicrosoftDocs` org which is pretty straight forward, [click step 3 here](https://review.docs.microsoft.com/en-us/help/get-started/setup-github?branch=main#3-join-microsoftdocs-and-other-organizations).  If you have any comment let's continue on private PR.

Actually, change is already live now https://docs.microsoft.com/en-us/azure/devops/artifacts/nuget/publish?view=azure-devops, so preview link is deleted. If you want to compare, then [previous version is here](https://review.docs.microsoft.com/en-us/azure/devops/artifacts/nuget/publish?branch=pazand-patch-1&view=azure-devops).

This PR show merged because private PR based on this, it was automatic step once private PR go live.
Let's continue discussion on private PR.

Thank you.

nkolev92 replied:
That doesn't seem to mentioned the authenticate task at all, is it on purpose?

erdembayar replied:
Authenticate task example is here [Azure Pipelines (YAML/Classic)](https://docs.microsoft.com/en-us/azure/devops/pipelines/artifacts/nuget?toc=%2Fazure%2Fdevops%2Fartifacts%2Ftoc.json&bc=%2Fazure%2Fdevops%2Fartifacts%2Fbreadcrumb%2Ftoc.json&view=azure-devops&tabs=yaml#publish-a-package) since it's related only to Azire pipeline. It's how they organize their docs.

erdembayar commented on docs/artifacts/includes/nuget/nuget-publish-endpoint.md:
```diff
@@ -18,14 +18,69 @@ ms.date: 02/18/2022
+
+### Pushing from other hosted builds
+
+When pushing from any other hosted builds (than local host, Azure hosted build, Github actions) to Azure Artifacts, your pipeline can use the [Azure PAT]([url](https://docs.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=Windows)) saved in `nuget.config` file for the authentication, the API key and UserName "AZ" are only us ...
```
Please look at [comment](https://github.com/MicrosoftDocs/azure-devops-docs/pull/12425#discussion_r906306644).

erdembayar commented on docs/artifacts/includes/nuget/nuget-publish-endpoint.md:
```diff
@@ -18,14 +18,69 @@ ms.date: 02/18/2022
+
+        :::image type="content" source="../../media/project-setup.png" alt-text="Project setup":::
+
+1. For Azure hosted builds that interact with Azure Artifacts feeds, you must use the [NuGet Authenticate](https://docs.microsoft.com/azure/devops/pipelines/tasks/package/nuget-authenticate) task for the authentication [(see examples)](#examples).
```
Please look at [comment](https://github.com/MicrosoftDocs/azure-devops-docs/pull/12425#discussion_r906306644).

erdembayar commented on docs/artifacts/includes/nuget/nuget-publish-endpoint.md:
```diff
@@ -18,14 +18,69 @@ ms.date: 02/18/2022
+
+1. For Github actions builds that interact with Azure Artifacts feeds, you may use [Github setup dotnet action](https://github.com/actions/setup-dotnet) for the authentication [(see examples)](#examples).
+
+1. For other hosted builds that interact with Azure Artifacts feeds, you may use [Azure PAT]([url](https://docs.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=Windows)) saved in `nuget.config` file for the authentication [(see examples)](#examples), but please exercise caution because it may leak your PAT as part of build  ...
```
Please look at [comment](https://github.com/MicrosoftDocs/azure-devops-docs/pull/12425#discussion_r906306644).
